vm/src/vm/runners/cairo_pie.rs

```diff
@@ -85,6 +87,35 @@ pub enum BuiltinAdditionalData {
     None,
 }
 
+impl BuiltinAdditionalData {
+    fn is_empty(&self) -> bool {
+        match self {
+            Self::Empty(_) => true,
+            Self::Hash(data) => data.is_empty(),
+            Self::Signature(data) => data.is_empty(),
+            Self::Output(_) => false,
+            Self::None => false,
+        }
+    }
+}
+
+impl PartialEq for BuiltinAdditionalData {
+    fn eq(&self, other: &BuiltinAdditionalData) -> bool {
+        if self.is_empty() | other.is_empty() {
+            // If either is empty, both have to be empty.
+            self.is_empty() & other.is_empty()
+        } else {
+            match (self, other) {
+                (Self::Hash(data), Self::Hash(other_data)) => data == other_data,
+                (Self::Signature(data), Self::Signature(other_data)) => data == other_data,
+                (Self::Output(data), Self::Output(other_data)) => data == other_data,
+                (Self::None, Self::None) => true,
+                _ => false,
+            }
+        }
```

Comment:
I think this should fix the error you found in my previous suggestion:
```suggestion
        match (self, other) {
            (Self::Hash(data), Self::Hash(other_data)) => data == other_data,
            (Self::Signature(data), Self::Signature(other_data)) => data == other_data,
            (Self::Output(data), Self::Output(other_data)) => data == other_data,
            (Self::None, Self::None) => true,
            (Self::Empty, x)|(x, Self::Empty) => x.is_empty(),
            _ => false,
        }
```
Emptiness for same variant instances is handled by the equality check, so we only need to ask that the other variant is empty if we know one is `Empty`. I'm not 100% sure it's legal as written due to the swapping order, worst case you need to write two arms instead.